plutus-core/untyped-plutus-core/src/Data/RandomAccessList/SkewBinary.hs

```diff
@@ -0,0 +1,105 @@
+{-# LANGUAGE BangPatterns    #-}
+{-# LANGUAGE LambdaCase      #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE TypeFamilies    #-}
+{-# LANGUAGE ViewPatterns    #-}
+module Data.RandomAccessList.SkewBinary ( RAList(Cons,Nil)
+                 , index
+                 , Data.RandomAccessList.SkewBinary.null
+                 , Data.RandomAccessList.SkewBinary.head
+                 , Data.RandomAccessList.SkewBinary.tail
+                 , uncons
+                 ) where
+
+import           Data.Bits  (unsafeShiftR)
+import qualified Data.List  as List (unfoldr)
+import           Data.Maybe
+import           GHC.Exts
+
+-- A complete binary tree.
+-- Note: the size of the tree is not stored/cached,
+-- unless it appears as a root tree in RAList, which the size is stored inside the Cons.
+data Tree a = Leaf a
+            | Node a !(Tree a) !(Tree a)
+            deriving (Eq, Show)
+
+-- a strict list of complete binary trees accompanied by their size.
+-- The trees appear in >=-size order.
+-- Note: this blist is strict on its spine, unlike the haskell's stdlib list
+data RAList a = BHead
+               {-# UNPACK #-} !Word -- ^ the size of the head tree
+               !(Tree a) -- ^ the head tree
+               !(RAList a) -- ^ the tail trees
+             | Nil
+             -- the derived Eq instance is correct,
+             -- because binary skew numbers have unique representation
```

Comment:
```suggestion
             -- because binary skew numbers have unique representation
             -- and hence all trees of the same size will have the same structure
```